### PR TITLE
TCP2: Simplify TCP packet creation

### PR DESCRIPTION
**Pull request by rveerama1**
There are existing calls to create net packet based on context, create IP header and finalize the packet before sending through net_send_data(). Test protocol also changed as well to verify with TTCN3.

### CODE REVIEW

File: subsys/net/ip/tcp2.c [hunk 1]
```diff
@@ -465,113 +466,122 @@ static size_t tcp_data_get(struct tcp *conn, struct net_pkt *pkt)
...
+	if (IS_ENABLED(CONFIG_NET_IPV6) && net_pkt_family(pkt) == AF_INET6) {
+		return net_ipv6_finalize(pkt, IPPROTO_TCP);
+	}
 
-	net_pkt_frag_insert(pkt, buf);
+	return -EINVAL;
 }
 
-static void ip_header_add(struct tcp *conn, struct net_pkt *pkt)
+static int tcp_header_add(struct tcp *conn, struct net_pkt *pkt, u8_t flags)
```
Comment (tbursztyka) [inline subsys/net/ip/tcp2.c:484]: call it tcp_create() to stay with the same naming we have all over the net stack.

File: subsys/net/ip/tcp2.c [hunk 2]
```diff
@@ -465,113 +466,122 @@ static size_t tcp_data_get(struct tcp *conn, struct net_pkt *pkt)
...
-		struct net_ipv6_hdr ip6;
 
-		memset(&ip6, 0, sizeof(ip6));
+	return net_pkt_set_data(pkt, &tcp_access);
+}
 
-		ip6.vtc = 0x60;
-		ip6.nexthdr = IPPROTO_TCP;
-		ip6.len = htons(net_pkt_get_len(pkt));
+static int ip_header_add(struct tcp *conn, struct net_pkt *pkt)
```
Comment (tbursztyka) [inline subsys/net/ip/tcp2.c:511]: hopefully this will go away once tcp2 will be properly integrated with net_context

File: subsys/net/ip/tp.c [hunk 3]
```diff
@@ -8,6 +8,8 @@
 #include <stdlib.h>
 #include "tp.h"
 #include "tp_priv.h"
+#include "ipv4.h"
```
Comment (tbursztyka) [inline subsys/net/ip/tp.c:11]: Could you split your patch in 2? tp is not tcp2.

File: subsys/net/ip/tp.c [hunk 4]
```diff
@@ -448,14 +424,36 @@ static void tp_pkt_send(struct net_pkt *pkt)
...
 void _tp_output(sa_family_t af, struct net_if *iface, void *data,
 		size_t data_len, const char *file, int line)
 {
-	struct net_pkt *pkt = tp_pkt_alloc(0, file, line);
+	struct net_pkt *pkt = tp_output_pkt_alloc(af, iface, file, line);
 	size_t total = data_len, len;
 	struct net_buf *buf;
 
-	pkt->family = af;
+	ip_header_add(pkt);
```
Comment (tbursztyka) [inline subsys/net/ip/tp.c:454]: why not using net_ipv4_create()/net_ipv6_create() ?

**Review by ozhuraki — APPROVED** [review]
Thanks and LGTM

File: subsys/net/ip/tp.c [hunk 5]
```diff
@@ -448,14 +424,36 @@ static void tp_pkt_send(struct net_pkt *pkt)
...
+
+	udp_header_add(pkt);
```
Comment (tbursztyka) [inline subsys/net/ip/tp.c:456]: same with net_udp_create()

File: subsys/net/ip/tp.c [hunk 6]
```diff
@@ -448,14 +424,36 @@ static void tp_pkt_send(struct net_pkt *pkt)
...
+	sys_slist_append(&tp_pkts, (sys_snode_t *) tp_pkt);
+
+	return tp_pkt->pkt;
+}
+
 void _tp_output(sa_family_t af, struct net_if *iface, void *data,
 		size_t data_len, const char *file, int line)
 {
-	struct net_pkt *pkt = tp_pkt_alloc(0, file, line);
+	struct net_pkt *pkt = tp_output_pkt_alloc(af, iface, file, line);
```
Comment (tbursztyka) [inline subsys/net/ip/tp.c:450]: you know the data length, go ahead: pass it to tp_output_pkt_alloc() and make it allocate for a udp hdr + the data length.

File: subsys/net/ip/tp.c [hunk 7]
```diff
@@ -465,14 +463,10 @@ void _tp_output(sa_family_t af, struct net_if *iface, void *data,
 
 		memcpy(net_buf_add(buf, len), data, len);
 
-		net_pkt_frag_add(pkt, buf);
+		net_pkt_append_buffer(pkt, buf);
```
Comment (tbursztyka) [inline subsys/net/ip/tp.c:466]: and then here get rid of the for () loop and us net_pkt_writg(pkt, data, len). Done.

**Review by tbursztyka — CHANGES_REQUESTED** [review]

File: subsys/net/ip/tcp2.c [hunk 8]
```diff
@@ -465,113 +466,122 @@ static size_t tcp_data_get(struct tcp *conn, struct net_pkt *pkt)
...
+{
+	if (IS_ENABLED(CONFIG_NET_IPV4) && net_pkt_family(pkt) == AF_INET) {
+		return net_context_create_ipv4_new(conn->context, pkt,
+						&(conn->src->sin.sin_addr),
```
Comment (jukkar) [inline subsys/net/ip/tcp2.c:515]: No need to put () after &, so `&conn->src->sin.sin_addr` is enough. Please fix all of them in this PR.

Comment (rveerama1) [reply subsys/net/ip/tcp2.c]: done

File: subsys/net/ip/tp.c (hunk 7 above)
Comment (rveerama1) [reply subsys/net/ip/tp.c]: done

File: subsys/net/ip/tp.c (hunk 5 above)
Comment (rveerama1) [reply subsys/net/ip/tp.c]: done

File: subsys/net/ip/tp.c (hunk 4 above)
Comment (rveerama1) [reply subsys/net/ip/tp.c]: ip_header_add does that. it calls net_ipv4_create and net_ipv6_create().

File: subsys/net/ip/tp.c (hunk 3 above)
Comment (rveerama1) [reply subsys/net/ip/tp.c:11]: done

File: subsys/net/ip/tcp2.c (hunk 2 above)
Comment (rveerama1) [reply subsys/net/ip/tcp2.c:511]: yep.

**Review by tbursztyka — APPROVED** [review]

**Review by pfl — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**